Sources/PackageExtension/PackageExtension.swift

```diff
@@ -0,0 +1,339 @@
+// to be emitted to SwiftPM verbatim. SwiftPM tries to interpret any stdout
+// contents after the last zero byte as a JSON encoded output struct in UTF-8
+// encoding; any failure to decode it is considered a protocol failure. The
+// exit code of the compiled extension determines success or failure (though
+// failure to decode the output is also considered a failure to run the ex-
+// tension).
+
+
+/// Provides information about the target being built, as well as contextual
+/// information such as the paths of the directories to which commands should
+/// be configured to write their outputs. This information should be used as
+/// part of generating the commands to be run during the build.
+public let targetBuildContext: TargetBuildContext = TargetBuildContextFactory()
+
+/// Constructs commands to run during the build, including full command lines.
+/// All paths should be based on the ones passed to the extension in the target
+/// build context.
+public let commandConstructor: CommandConstructor = CommandConstructor()
+
+/// Emits errors, warnings, and remarks to be shown as a result of running the
+/// extension. After emitting one or more errors, the extension should return a
+/// non-zero exit code.
+public let diagnosticsEmitter = DiagnosticsEmitter()
+
+
+/// Provides information about the target being built, as well as contextual
+/// information such as the paths of the directories to which commands should
+/// be configured to write their outputs. This information should be used as
+/// part of generating the commands to be run during the build.
+public final class TargetBuildContext: Decodable {
+
+    /// The name of the target being built.
+    public let targetName: String
+    
+    /// The module name of the target. This is usually derived from the name,
+    /// but is possibly be customizable in the package manifest in some future
+    /// SwiftPM version).
+    public let moduleName: String
+    
+    /// The path of the target directory.
+    public let targetDir: Path
+    
+    /// That path of the package that contains the target.
+    public let packageDir: Path
+    
+    /// Absolute paths of the source files in the target. This might include
+    /// derived source files generated by other extensions).
+    public let sourceFiles: [Path]
+
+    /// Absolute paths of the resource files in the target.
+    public let resourceFiles: [Path]
+
+    /// Absolute paths of any other files (not sources or resources) in the target.
+    public let otherFiles: [Path]
+
+    /// Information about all targets on which the target for which the exten-
+    /// sion is being invoked either directly or indirectly depends. This list
+    /// is in topologically sorted order, with immediate dependencies appearing
+    /// earlier and more distant dependencies later in the list. This is mainly
+    /// intended for generating lists of search path arguments, etc.
+    public let dependencies: [DependencyTargetInfo]
```

Comment:
Only some tools, such as `protoc`, need to see information about the dependency closure.